### ISSUE TITLE
storeref expressions

### ISSUE THREAD

**Issue by davidcok** [JavaModelingLanguage/RefMan#15]
The phrase store-ref expression has always seemed odd to me. It doesn't really connect to other PL literature.
l-value expression would be more intuitive (to me).

I'd prefer to talk about *program state*, which consists of a set of (memory) *locations*, which are referred to by *l-value expressions*.

But I'd be interested in knowing what terminology is actually used in teaching these concepts.

**Comment by wadoon** [comment]
I have no opinion against store-ref. I find it very intuitive as an abbreviation of storage/reference-expression. 

On the other side, l-value and r-value are commonly used in C++, but (at least in my courses) they were not prominent in the education of the Java language. Also the analogy breaks if we allow `o.*` and `o[*]` which are definitiv not a lhs expression. 

I have no obstacle against a re-labeling of the store-refs, e.g., to memory location (set) expression, but I would avoid calling them l-values excessively.

**Comment by leavens** [comment]
Hi Alexander, David, and all,

I agree. It's a good point that o.I and o[*] cannot be used on the left-hand side of an assignment.


        Regards,

        Gary T. Leavens
        329 L3Harris Center
        Computer Science, University of Central Florida
        4000 Central Florida Blvd., Orlando, FL 32816-2362 USA
        http://www.cs.ucf.edu/~leavens phone: +1-407-823-4758
        ***@***.******@***.***>

**Comment by davidcok** [comment]
True, but they stand for sets of l-values.

So Gary, are you saying you want to stick with storeref expression?

**Comment by leavens** [comment]
Hi David and all,

Yes, I think the term "store-ref expression" is fine, and it's somewhat traditional in language references to coin new terms for notions that don't precisely match old terms.


        Regards,

        Gary T. Leavens
        329 L3Harris Center
        Computer Science, University of Central Florida
        4000 Central Florida Blvd., Orlando, FL 32816-2362 USA
        http://www.cs.ucf.edu/~leavens phone: +1-407-823-4758
        ***@***.******@***.***>

**Comment by mattulbrich** [comment]
I am also in favor of not using l-value. From my KeY perspective, I'd say there are memory locations, location set expressions (with their own type \locset) and location expressions (referencing to a single location). storerefs is the unifying notion for the syntactical units.

**Comment by davidcok** [comment]
OK. I'll go with

(program) state -- the set of all (memory) locations (fields and array elements of classes and reachable object references)

location set expressions (type \locset) -- expressions denoting some number of locations, including the literals below and terms like \set_union (and others)

no way to express a single element of a \locset -- always at least a singleton \locset

storeref -- syntactical means to express \locset literals: x o.f o.* a[1] a[1..2] a[ * ] \nothing \everything

**Comment by leavens** [comment]
David's last comment is all fine with me.

**Comment by mattulbrich** [comment]
Just to verify: Locations are memory places on the heap, not on the stack. A local variable is not a location in KeY-speak.

Do you agree?

**Comment by leavens** [comment]
No, I would say that a location can also be on the stack, but your terminology is often used in the literature.  For an assignable clause, we want to specify both locations on the heap and on the stack.

**Comment by mattulbrich** [comment]
Now that is an interesting subtlety.

It is easier to model location sets as "heap-only". So I'd suggest they remain such.
Perhaps that is a difference between location sets and storeref in that storeref can also refer to local variables. Since location is a new term in JML, we could make it refer to heap locations.

(Local variables are only relevant inside methods and thus actually never used in dynamic frames imho. They may occur in local assignable clauses, but KeY infers them there automatically.)

**Comment by davidcok** [comment]
Now that we have statement (block) specs, I thought we needed to have local variables in location sets.
The block spec summarizes the action of a bock of code, so the summary has to state what stack variables might have been assigned as well as heap locations. As long as we only had method specs this was not an issue.

**Comment by leavens** [comment]
I agree with David's thought on this.

**Comment by mattulbrich** [comment]
We need them in storerefs, yes, but not in location sets, I'd say. At least that's KeY's representation. We would have problems in having a \locset variable containing a local variable.

On major problem is that local variables have a limited lifetime, but may endure after their expiring in a location set, yield a situation of "dangling reference". Moreover, there is the problem that a local variable can exist multiple times:

````
class C {
   //@ ghost \locset l;

   void m(int n) {
      //@ set l = l + \locset(x);    // nevermind the concrete syntax for a singleton location set.
      if(n > 0) {
         m(n-1);
      }
   }

   void main() {
      //@ set l = \locset();  // or \set_empty to speak KeY.
      m(10);
      //@ assert \cardinality(l) == 10; // or
      //@ assert \cardinality(l) == 1;                 // (cardinality is not defined in JML, but you get its idea.)
   }
}
````

**Comment by wadoon** [comment]
Before we forget, here is a special behavior on storerefs in KeY. 

Consider the following example:

```
class A {
  ... 
  //@ private \locset footprint = \locset(this.a);
  //@ assignable footprint;
  public void foo() {}
} 
```

If a variable occurring in a storeref expression is of type `\locset`, then the content of this variable is added to the storeref and not the variable itself. In the example, the `assignable` clause includes `this.a` and not `this.footprint`. If you want to add `footprint` to the assignable heap locations, you need to "quote" it by using `\locset(footprint)`.

**Comment by leavens** [comment]
I think we may have a consensus on this. I agree with Alexander's comment about needing to "quote" footprints to prevent them from being expanded if you want to refer to the ghost field that holds a location set.

**Comment by davidcok** [comment]
The original post was about terminology -- on that there is a consensus.

We've morphed to a discussion of local variables in locsets. As I see it,
- in block contracts, assignable clauses need to be able to refer to in-scope local variables
- if locsets are a stand-in/desugaring for assignable clauses, local variables need to be a part of locsets as well, otherwise an assignable clause is not quite an equivalent to a locset expression
- but this opens up the possibility of exporting a locset containing a scope-restricted location outside of that scope
- (lambda expression rules constrain capturing/exporting local variables by requiring them to be "effectively final", by which I presume just the value is used)

I see two possible resolutions:
a) allow local variables in assignable clauses (they can only be simple names), but do not allow them in locsets, so they have to be treated separately from non-local variables
b) find a way to disallow having local variables in locsets be captured outside their scopes (that is, assigned to variables/fields outside their scope)
To me (a) seems less aesthetically pleasing but simpler, and it is my understanding of what KeY does. And even if Key infers the local variables that are assigned, it still is appropriate to be able to write down explicitly what that inference is.
For example, in a block contract
```
//@ refining
//@ assignable ...
{
... very long block...
}
```
Where the block contract is seen as a summary of the actions of the block, one wants very much for the assignable clause to list what local variables (in scope outside the block) are written to by the block.

Note though that array elements and object fields of local variables are in the heap and can very much be altered and inspected outside the scope of the local variable that originally held them. Consider the following example (you can do the same for array elements).
```
public class Cap {

int k = 42;

public static java.util.function.Consumer<Integer> c;
public static java.util.function.Supplier<Integer> s;

public static void m() {
  final Cap localcap = new Cap();
  c = i->{localcap.k = 10 + i;};
  s = ()->localcap.k;
}

public static void main(String... args) {
  m();
  c.accept(20);
  System.out.println(s.get());
  c.accept(100);
  System.out.println(s.get());
}
}
```
There is no way to refer the to Cap object that contains the k field being altered and read outside of m(), even  though it still exists on the heap. To specify this I think one would need to declare a ghost variable (say `ghostcap`)that holds the value of `localcap` in order to specify what c.accept() modifies and s.get() reads. And then there would be a locset containing `ghostcap.k`

**Comment by mattulbrich** [comment]
>I see two possible resolutions:
a) allow local variables in assignable clauses (they can only be simple names), but do not allow them in locsets, so they have to be treated separately from non-local variables
b) find a way to disallow having local variables in locsets be captured outside their scopes (that is, assigned to variables/fields outside their scope)
To me (a) seems less aesthetically pleasing but simpler, and it is my understanding of what KeY does. And even if Key infers the local variables that are assigned, it still is appropriate to be able to write down explicitly what that inference is.

I quite like (a). Locations are memory places on the heap, that is where the whole framing business takes place, only there can be aliasing etc. The stack is harmless, so it is not in my mental model of locations, and thus the situation does not displease me particularly.

the expressions in assignable clauses would indeed be locations and local variables. But that's all I think. Our student swallow that usually quite happily.

**Comment by leavens** [comment]
I agree to (a) also, let the local variables be named explicitly.

**Comment by mattulbrich** [comment]
> And even if Key infers the local variables that are assigned, it still is appropriate to be able to write down explicitly what that inference is.

And in KeY, you can write that down. It is currently bluntly ignored, so you can write a super/subset of the modified variables, nobody cares. There might be a warning in order.
